src/Material/Tooltip.elm

```diff
@@ -0,0 +1,523 @@
+> visual element that displays related but typically non-essential content,
+> e.g., a definition, clarification, or brief instruction.
+>
+> Tooltips are a ubiquitous feature of most user interfaces, regardless of a
+> site's content or function. Their design and use is an important factor in the
+> overall user experience. See the tooltip component's Material Design
+> specifications page for details.
+
+See also the
+[Material Design Specification](https://material.google.com/components/tooltips.html).
+
+Refer to [this site](http://debois.github.io/elm-mdl#/tooltips)
+for a live demo.
+
+
+# Types
+@docs Model, defaultModel
+@docs DOMState
+@docs Property
+
+# Update and render
+@docs Msg, update, view, render
+
+# Content
+@docs left, right, top, bottom
+@docs default, large
+
+# Events
+@docs onMouseEnter, onMouseLeave
+@docs onEnter, onLeave
+@docs attach
+
+#Utility
+@docs mdl
+
+-}
+
+import Platform.Cmd exposing (Cmd, none)
+import Html exposing (..)
+import Parts exposing (Indexed)
+import Material.Options as Options exposing (Style, cs, css, when)
+import Material.Options.Internal as Internal
+import DOM
+import Html.Events
+import Json.Decode as Json exposing ((:=), at)
+import String
+
+
+-- MODEL
+
+
+{-| Component model.
+-}
+type alias Model =
+  { isActive : Bool
+  , domState : DOMState
+  }
+
+
+{-| Default component model constructor.
+-}
+defaultModel : Model
+defaultModel =
+  { isActive = False
+  , domState = defaultDOMState
+  }
+
+
+
+-- ACTION, UPDATE
+
+
+{-| Component action.
+-}
+type Msg
+  = Enter DOMState
+  | Leave
+
+
+{-| Tooltip position
+-}
+type alias Pos =
+  { left : Float
+  , top : Float
+  , marginLeft : Float
+  , marginTop : Float
+  }
+
+
+{-| Default position constructor
+-}
+defaultPos : Pos
+defaultPos =
+  { left = 0
+  , top = 0
+  , marginLeft = 0
+  , marginTop = 0
+  }
+
+
+{-| Position and offsets from dom events for the tooltip
+-}
+type alias DOMState =
+  { rect : DOM.Rectangle
+  , offsetWidth : Float
+  , offsetHeight : Float
+  }
+
+
+{-| Default DOMState constructor
+-}
+defaultDOMState : DOMState
+defaultDOMState =
+  { rect = { left = 0, top = 0, width = 0, height = 0 }
+  , offsetWidth = 0
+  , offsetHeight = 0
+  }
+
+
+{-| Calculates the position of the tooltip based on the event
+and the requested position
+-}
+calculatePos : Position -> DOMState -> Pos
+calculatePos pos domState =
+  let
+    -- _ = Debug.log "Calculating position for " domState
+    props =
+      domState.rect
+
+    offsetWidth =
+      domState.offsetWidth
+
+    offsetHeight =
+      domState.offsetHeight
+
+    left =
+      props.left + (props.width / 2)
+
+    top =
+      props.top + (props.height / 2)
+
+    marginLeft =
+      -1 * (offsetWidth / 2)
+
+    marginTop =
+      -1 * (offsetHeight / 2)
+
+    out =
+      case pos of
+        Left ->
+          { left = props.left - offsetWidth - 10
+          , top =
+              if (top + marginTop < 0) then

```

Comment:
Very difficult to read. Maybe write the `if` on a single line:

    , top = if (top + marginTop < 0) then 0 else top

Same multiple times below.